data-prepper-api/src/main/java/com/amazon/dataprepper/model/event/JacksonEventImpl.java

```diff
@@ -0,0 +1,252 @@
+    private final EventMetadata eventMetadata;
+
+    private final JsonNode jsonNode;
+
+    private final ObjectMapper mapper;
+
+    public JacksonEventImpl(final Builder builder) {
+
+        this.mapper = new ObjectMapper();
+
+        if (builder.eventMetadata == null) {
+            this.eventMetadata = new EventMetadataImpl.Builder()
+                    .withEventType(builder.eventType)
+                    .withTimeReceived(builder.timeReceived)
+                    .withAttributes(builder.eventMetadataAttributes)
+                    .build();
+        } else {
+            this.eventMetadata = builder.eventMetadata;
+        }
+
+        if (builder.data == null) {
+            this.jsonNode = mapper.valueToTree(new HashMap<>());
+        } else {
+            this.jsonNode = mapper.valueToTree(builder.data);
+        }
+
+    }
+
+    /**
+     * Adds or updates the key with a given value in the Event.
+     * @param key where the value will be set
+     * @param value value to set the key to
+     * @since 1.2
+     */
+    @Override public void put(final String key, final Object value) {
+
+        checkKeyArgument(key);
+
+        final LinkedList<String> keys = new LinkedList<>(Arrays.asList(key.split(DOT_NOTATION_REGEX)));
+        recursivelyTraverseAndPut(jsonNode, keys, value);
+    }
+
+    private void recursivelyTraverseAndPut(final JsonNode node, final LinkedList<String> keys, final Object value) {
+
+        if (keys.size() == 1) {
+            final JsonNode valueNode = mapper.valueToTree(value);
+            ((ObjectNode) node).set(keys.get(0), valueNode);
+        } else {
+            final String key = keys.removeFirst();
+            JsonNode baseNode = node.get(toJsonPointerExpression(key));
+            if (baseNode == null) {
+                baseNode = mapper.createObjectNode();
+                ((ObjectNode) node).set(key, baseNode);
+            }
+            recursivelyTraverseAndPut(baseNode, keys, value);
+        }
+    }
+
+    /**
+     * Retrieves the value of type clazz from the key.
+     * @param key the value to retrieve from
+     * @param clazz the return type of the value
+     * @return the value
+     * @throws RuntimeException if it is unable to map the value to the provided clazz
+     * @since 1.2
+     */
+    @Override public <T> T get(final String key, final Class<T> clazz) {
+
+        checkKeyArgument(key);
+        final String jsonPointerExpression = getBaseJsonPointerExpression(key);
+
+        JsonNode node = jsonNode.at(jsonPointerExpression);
+        if (node.isMissingNode()) {
+            return null;
+        }
+
+        try {
+            return mapper.treeToValue(node, clazz);
+        } catch (final JsonProcessingException e) {
+            LOG.error("Unable to map {} to {}", key, clazz, e);
+            throw new RuntimeException(String.format("Unable to map %s to %s", key, clazz), e);
+        }
+    }
+
+    private String toJsonPointerExpression(final String key) {
```

Comment:
It might be nice to use the `JsonPointer` class here so it is quite clear what it is returning. But, these are private, so it's not critical.